### DISCUSSION TITLE
Guidance for new persons to scripting/development in FP?

### DISCUSSION THREAD

**Discussion by dimmech** [freeplane/freeplane#1575] [Help] [answered]
Hello, I've been using FP for a short while and am interested in how I can jump into the scripting. However, I'm running into issues out of the gate. Using version 1.11.8 in windows I'm trying to install scriptlib and running into errors on restart. First, I get duplicate class definition errors because of the files in the "src" directory that remain after the installation. After deleting those I get "unable to resolve class" errors on the 4 groovy files included in the package.

I'm trying to go through FP documentation and understand it's a work in progress and have also done some searching here in the forums on the states of some of these plugins, whether or not they're outdated, if there deprecated, or rather, if I'm doing something else wrong.  I'm looking for some exercises that deliver the basic concepts in simpler steps. My apologies if I missed something obvious.

**Comment by i-plasm** [comment, marked as answer]
Hello. I think the add-on you mentioned is out-of-date (at least, I suspect that since the sourceforge version is very old). Freeplane v1.10.x introduced major changes in groovy and java versions. Freeplane's API has changed as well throughout the years. I advice you follow a sequence similar to this for your specific case:
1. I would not recommend using any third party libs at this early stage, in order to avoid running into possible incompatibilities. So, ignore or delete plug-ins which might be outdated.
1. Get the basics of scripting: https://docs.freeplane.org/scripting/Scripting.html
1. Practice the basics by studying and modifying some of the scripts here: https://docs.freeplane.org/scripting/Scripts_collection.html . (I'm not sure all of them are fully working, but I believe most of them are).
1. Once you have covered the basics, dive into the API https://docs.freeplane.org/api/index-all.html and apply it on existing or new scripts.
1. When you feel confident in scripting, you can consider creating plug-ins.
1. NOTE: it is important to be mindful of scripting preferences ( ` Tools -> Preferences -> Plugins -> Scripting (section) ` ). There, script permissions are to be set (i.e read, write, etc). Different scripts may require different permissions. Additionally, the script and library classpath can be modified and extended in those preferences. Some issues sometimes have to do with improper setting of those preferences.

Regarding the IDE: I use Eclipse. If it interests you, there are instructions on IntelliJ Idea. A simplified set up is detailed here: https://github.com/freeplane/freeplane/discussions/589#discussioncomment-3472452. A full set up is detailed here: https://github.com/freeplane/freeplane/discussions/1321#discussioncomment-6631241  . I recommend initially a simplified set up, and once you become comfortable with scripting, you can go for the full developing environment (either in Eclipse or IntelliJ Idea). I haven't seen instructions for other environments. Obviously you can use any plain text editor for scripting, but it is not as convenient as using a properly set up IDE.

An extra recommendation: when scripting, if you encounter an error that has to do with _unnamed module of loader_, which reads like ` java.security.PrivilegedActionException: java.lang.ClassCastException .... unnamed module of loader org.knopflerfish.framework.BundleClassLoader `, try deleting the `compiledscripts` and `compiledscripts2` folders from your user directory. You can find that directory at: ` Tools -> Open user directory ` . In my experience, at least, when I've encountered such error it is fixed by deleting those folders - maybe it's due to a freeplane bug -. Those folders will be automatically generated with fresh and new compilations once the need arises. For changes to become effective, a Freeplane restart is required.

Make sure to only execute scripts that are safe and reliable.
I hope this helps!
Regards

**Reply by dimmech** [reply]
Excellent, I appreciate your response. This "sequence" and your other tips are exactly what I was looking for, somewhat of a quick start guide. Thanks!
